--- linux/drivers/net/wireless/zd1211rw/zd_mac.c.orig	2007-04-01 02:54:33.000000000 +0200
+++ linux/drivers/net/wireless/zd1211rw/zd_mac.c	2007-04-01 03:08:23.000000000 +0200
@@ -137,18 +137,9 @@
 static int reset_mode(struct zd_mac *mac)
 {
 	struct ieee80211_device *ieee = zd_mac_to_ieee80211(mac);
-	struct zd_ioreq32 ioreqs[3] = {
-		{ CR_RX_FILTER, STA_RX_FILTER },
-		{ CR_SNIFFER_ON, 0U },
-	};
-
-	if (ieee->iw_mode == IW_MODE_MONITOR) {
-		ioreqs[0].value = 0xffffffff;
-		ioreqs[1].value = 0x1;
-		ioreqs[2].value = ENC_SNIFFER;
-	}
-
-	return zd_iowrite32a(&mac->chip, ioreqs, 3);
+	u32 filter = ieee->iw_mode == IW_MODE_MONITOR ? ~0 : STA_RX_FILTER;
+ 
+	return zd_iowrite32(&mac->chip, CR_RX_FILTER, filter);
 }
 
 int zd_mac_open(struct net_device *netdev)
@@ -834,7 +825,8 @@
 	const struct rx_status *status;
 
 	*pstatus = status = zd_tail(buffer, length, sizeof(struct rx_status));
-	if (status->frame_status & ZD_RX_ERROR) {
+	if (status->frame_status & ZD_RX_ERROR
+		|| status->frame_status & ~0x21) {
 		/* FIXME: update? */
 		return -EINVAL;
 	}
