diff -ur madwifi-ng-r1545/ath/if_ath.c patched_madwifi-ng-r1545/ath/if_ath.c
--- madwifi-ng-r1545/ath/if_ath.c	2006-05-05 22:20:22.000000000 +0200
+++ patched_madwifi-ng-r1545/ath/if_ath.c	2006-05-10 19:02:00.000000000 +0200
@@ -2246,6 +2246,7 @@
 ath_tx_startraw(struct net_device *dev, struct ath_buf *bf, struct sk_buff *skb) 
 {
 	struct ath_softc *sc = dev->priv;
+	struct ieee80211com *ic = &sc->sc_ic;
 	struct ath_hal *ah = sc->sc_ah;
 	struct ieee80211_phy_params *ph = (struct ieee80211_phy_params *) (skb->cb + sizeof(struct ieee80211_cb));
 	const HAL_RATE_TABLE *rt;
@@ -2262,7 +2263,11 @@
 	struct ieee80211_frame *wh; 
 	
 	wh = (struct ieee80211_frame *) skb->data;
-	try0 = ph->try0;
+	try0 = (ic->ic_opmode == IEEE80211_M_MONITOR) ? 1 : ph->try0;
+	/*
+	 * The retry value has to be patched to 1 when injecting,
+	 * otherwise the sequence number will be overwritten
+	 */
 	rt = sc->sc_currates;
 	txrate = dot11_to_ratecode(sc, rt, ph->rate0);
 	power = ph->power > 60 ? 60 : ph->power;
